presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarFilterScanToJoin.java

```diff
@@ -0,0 +1,124 @@
+ * </pre>
+ * <p>
+ * This optimizer can rewrite to aggregation over a left outer join:
+ * <p>
+ * From:
+ * <pre>
+ * - LateralJoin (with correlation list: [C])
+ *   - (input) plan which produces symbols: [A, B, C]
+ *   - (subquery) Project F
+ *     - Filter(D = C AND E > 5)
+ *       - plan which produces symbols: [D, E, F]
+ * </pre>
+ * to:
+ * <pre>
+ * - Filter(CASE CN WHEN 0 true WHEN 1 true ELSE fail('Scalar sub-query has returned multiple rows'))
+ *   - Aggregation(GROUP BY A, B, C, U; functions: [count(non_null) AS CN, arbitrary(F1) AS F...]
+ *     - Join(LEFT_OUTER, D = C)
+ *       - AssignUniqueId(adds symbol U)
+ *         - (input) plan which produces symbols: [A, B, C]
+ *       - Filter(E > 5)
+ *         - projection which adds no null symbol used for count() function
+ *           - plan which produces symbols: [D, E, F1]
+ * </pre>
+ * <p>
+ * Note only conjunction predicates in FilterNode are supported
+ */
+public class TransformCorrelatedScalarFilterScanToJoin
+        implements Rule
+{
+    private static final Pattern PATTERN = Pattern.node(LateralJoinNode.class);
+    private final FunctionRegistry functionRegistry;
+
+    public TransformCorrelatedScalarFilterScanToJoin(FunctionRegistry functionRegistry)
+    {
+        this.functionRegistry = requireNonNull(functionRegistry, "functionRegistry is null");
+    }
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        LateralJoinNode lateralJoinNode = (LateralJoinNode) node;
+        PlanNode subquery = lookup.resolve(lateralJoinNode.getSubquery());
+
+        if (lateralJoinNode.getCorrelation().isEmpty() || !isScalar(subquery, lookup)) {
+            return Optional.empty();
+        }
+
+        Optional<FilterNode> filterScan = getFilterScan(subquery, lookup);
+        if (!filterScan.isPresent()) {
+            return Optional.empty();
+        }
+
+        ScalarSubqueryToJoinRewriter rewriter = new ScalarSubqueryToJoinRewriter(functionRegistry, symbolAllocator, idAllocator, lookup);
+        return rewriter.rewriteScalarFilterScan(lateralJoinNode, filterScan.get());
+    }
+
+    private Optional<FilterNode> getFilterScan(PlanNode planNode, Lookup lookup)
+    {
+        if (!searchFrom(planNode, lookup)
+                .where(TableScanNode.class::isInstance)
+                .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, FilterNode.class, EnforceSingleRowNode.class))
+                .matches()) {
+            return Optional.empty();
+        }
+
+        List<FilterNode> filterNodes = searchFrom(planNode, lookup)
+                .where(FilterNode.class::isInstance)
+                .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, EnforceSingleRowNode.class))
+                .findAll();
+
+        if (filterNodes.size() != 1) {
```

Comment:
I'm thinking about cases where filterNodes would be > 1. Do we ever do something like:
```
Filter: (x+1 > 3)
     Project: output: ( x+1, y)
             Filter(y < 7)
                    TableScan: output: x,y
```
Because in that case it would be valuable to pull the predicate up to combine the filter nodes so we could apply this rewrite (not suggesting we do that now, just thinking about limitations)